perf/src/recycler.rs

```diff
@@ -73,41 +181,132 @@ impl<T: Default + Reset + Sized> Recycler<T> {
         new
     }
 
-    pub fn allocate(&self, name: &'static str) -> T {
-        let new = self
-            .recycler
-            .gc
-            .lock()
-            .expect("recycler lock in pb fn allocate")
-            .pop();
-
-        if let Some(mut x) = new {
-            self.recycler.stats.reuse.fetch_add(1, Ordering::Relaxed);
-            x.reset();
-            return x;
-        }
-
-        let total = self.recycler.stats.total.fetch_add(1, Ordering::Relaxed);
-        trace!(
-            "allocating new: total {} {:?} id: {} reuse: {} max_gc: {}",
-            total,
-            name,
-            self.recycler.id,
-            self.recycler.stats.reuse.load(Ordering::Relaxed),
-            self.recycler.stats.max_gc.load(Ordering::Relaxed),
-        );
+    pub fn allocate(&self) -> Option<T> {
+        let mut shrink_removed_objects = vec![];
+        let (mut allocated_object, did_reuse, should_allocate_new, mut shrink_stats) = {
+            let mut object_pool = self
+                .recycler
+                .gc
+                .lock()
+                .expect("recycler lock in pb fn allocate");
+
+            let now = timestamp();
```

Comment:
I think it's generally more suited to use `Instant` just directly rather than `timestamp()` (which is backed by `SystemTime`). Because of monotonically non-decreasing nature of `Instant` over `SystemTime`.